jax/experimental/tf_bridge/jax_to_tf.py

```diff
@@ -0,0 +1,841 @@
+  """Shape inference function for reduce_window_{sum,min,max}."""
+  params = dict(window_dimensions=window_dimensions,
+                window_strides=window_strides,
+                padding=padding,
+                input_shape=input_shape)
+  try:
+    out, = _infer_shape_jax(jax_f, operand, **params)
+  except TypeError:
+    del params["input_shape"]
+    out, = _infer_shape_jax(jax_f, operand, **params)
+  return out.shape
+
+
+def _get_shape_from_tensor_or_array(x):
+  if isinstance(x.shape, tf.TensorShape):
+    return tuple(x.shape.as_list())
+  return tuple(x.shape)
+
+
+def _reduce_window(jax_f, reducer, init_val, operand, window_dimensions,
+                   window_strides, padding, input_shape=None):
+  """TensorFlow implementation of reduce_window_{sum,min,max}."""
+  del input_shape
+  # TODO(tomhennigan): tf2xla should have a shape inference function.
+  out_shape = _reduce_window_shape(jax_f, operand, window_dimensions,
+                                   window_strides, padding)
+  padding = lax.padtype_to_pads(_get_shape_from_tensor_or_array(operand),
+                                window_dimensions,
+                                window_strides, padding)
+  a = tf.constant(0, operand.dtype)
+  reducer_fn = reducer.get_concrete_function(a, a)
+  out = tfxla.reduce_window(operand, tf.constant(init_val, operand.dtype),
+                            reducer_fn, window_dimensions,
+                            window_strides, padding=padding)
+  out.set_shape(out_shape)
+  return out
+# pylint: disable=protected-access
+tf_impl[lax.reduce_window_sum_p] = (
+    functools.partial(_reduce_window, lax._reduce_window_sum, _add_fn, 0))
+tf_impl[lax.reduce_window_min_p] = (
+    functools.partial(_reduce_window, lax._reduce_window_min, _min_fn, np.inf))
+tf_impl[lax.reduce_window_max_p] = (
+    functools.partial(_reduce_window, lax._reduce_window_max, _max_fn, -np.inf))
+# pylint: enable=protected-access
+
+
+def _select_and_scatter_add(
+    operand, source, init_value, select_jaxpr, select_consts, scatter_jaxpr,
+    scatter_consts, window_dimensions, window_strides, padding):
+  del select_jaxpr, select_consts, scatter_jaxpr, scatter_consts
+  # TODO(phawkins): handle the select and scatter jaxprs correctly.
+  a = tf.constant(0, operand.dtype)
+  select_fn = _ge_fn.get_concrete_function(a, a)
+  scatter_fn = _add_fn.get_concrete_function(a, a)
+  return tfxla.select_and_scatter(operand, window_dimensions, window_strides,
+                                  padding, source, init_value, select_fn,
+                                  scatter_fn)
+tf_impl[lax.select_and_scatter_add_p] = _select_and_scatter_add
+
+
+def uadd(a, *b):
+  """Workaround to support + with uint32 (not supported in TF)."""
+  # Note: Tensorflow's add_n doesn't support broadcasting.
+  b = [tf.broadcast_to(b, tf.shape(a)) for b in b]
+  return tf.add_n([a] + b)
+
+
+def _threefry2x32(key1, key2, x1, x2):
```

Comment:
I realize this isn't new in this PR most likely, but I'm a bit unhappy that we have to repeat the definition of threefry here. Note that on CPU we don't have a direct definition of the primitive; we expand it using `xla.lower_fun`. Could we do something similar here rather than repeating its definition?

Reply:
Added a TODO, will do separately.